Support type-constrained destructuring

danielwe opened the pull request:
Fixes #54 

To manage the combinatorial explosion of composite argument expressions, I made `inject_symbol_to_arg` recursive, which means it now also supports `f((; x)::Foo...=foo) = ...` and other deeply cursed incantations. I didn't implement tests for such combinations, however.

MilesCranmer commented:
I think the `(; x) = ...` syntax doesn't work on Julia 1.6-?, so maybe the new test can be fenced off for those versions

MilesCranmer commented on test/unittests.jl:
```diff
@@ -52,6 +52,25 @@ end
+        @eval @stable default_codegen_level = $codegen_level f((x,)::Vector) = x
+        @test f([1]) == 1
+        @test_throws MethodError f((1,)) == 1
+        @eval @stable default_codegen_level = $codegen_level g((; x)::A) = x
+        @test g(B(1)) == 1
+        @test_throws MethodError g((; x=1))
```
Could you add some kind of test here for the expected generated code, with the exact destructuring syntax expected? e.g., you could just have some kind of `occursin("...", string(expr))` and feed it the output of `@macroexpand`.

MilesCranmer commented on test/unittests.jl:
```diff
@@ -52,6 +52,25 @@ end
+        @eval @stable default_codegen_level = $codegen_level f((x,)::Vector) = x
+        @test f([1]) == 1
+        @test_throws MethodError f((1,)) == 1
+        @eval @stable default_codegen_level = $codegen_level g((; x)::A) = x
```
To strengthen the tests, could you also add cases for:

1. Multiple keywords in the destructuring, like `g((; x, y))`
2. Multiple instances of destructuring in the same signature, like `g((; a, b), (; c))`
3. Mixed kwarg and tuple destructuring, like `g((x, y), (; a, b))`

MilesCranmer commented:
(Also, huge thanks for adding this!!)

MilesCranmer commented:
Does this also account for functions like the following?

```julia
julia> function f(x, (; a, b) = (; a=x, b=2, c=3))
           a + b
       end
f (generic function with 2 methods)

julia> f(1)
3
```

Where as can be seen, the value of `x` is used in the instantiation of `a`

danielwe commented:
Thanks for the feedback. I added a bunch of tests addressing the points you mentioned, including your last example, which does indeed work (what's happening here is more or less the same as what the lowering pass does anyway, and it only touches the left-hand side of `::`).

Let me know if there's anything else you want to see. (For example, style-wise, there are some very long lines, but I wasn't sure what your preferred way of dealing with them would be.)

MilesCranmer commented:
Thanks! Looking good. One quick style comment is I wonder if the tests could be broken down into additional `@testitem`s, to give clearer test reports? e.g., one test item on (; x), one on (x,), etc. It helps with maintainability/refactoring, since it’s easy to run specific tests in VSCode, and also once TestItemRunner parallelises things, it means smaller testitems can be run in parallel (though not too small to take up too much of the unit test logging info).

MilesCranmer commented:
Also, I just confirmed it doesn't re-trigger #51 which is good. (It's a Julia bug in 1.11 but we can't figure out what really triggers it)

MilesCranmer commented:
Just FYI I'm going to make some edits to the PR and push, so let me know if you have any uncommitted changes. I think it's ready to merge after those edits though.

danielwe commented:
Go ahead! I was going to get to it later today, but if you've already taken care of it then all the better!

danielwe commented on src/stabilization.jl:
```diff
@@ -227,12 +227,14 @@ function _stabilize_fnc(
-    args = map(inject_symbol_to_arg, func[:args])
+    args, destructurings = let
+        args_destructurings = map(inject_symbol_to_arg, func[:args])
+        map(first, args_destructurings), map(last, args_destructurings)
```
Note that not filtering out the `nothing`s here will lead to several lines with just the single statement `nothing` in the function body. Of course, these will be compiled away, but still pollute `@code_warntype` etc.

MilesCranmer replied:
Good point, I changed it to just be `Expr(:block)`

MilesCranmer replied:
Actually I'll just change it back. Seems like it could result in a lot of extra work for the compiler in cases where there are a lot of args

MilesCranmer commented:
Ok, I think that should do it. Let me know if you have any other comments!

MilesCranmer commented:
The slight reduction in coverage seems to be from the catch-all

```julia
    return ex, nothing
```

I'm not sure what scenarios would trigger this, so we could probably leave this uncovered for now until we think of one.

MilesCranmer reviewed (approved):
Thanks again for the great PR!
